### PR TITLE
Closes #308 - Use uuids instead of database IDs.

### LINKED ISSUE #308: Use UUIDs instead of primary keys
We don't expose primary keys (`id`) anymore, but either way, `/games/3` should not look up `SELECT slug, game_url, name, description FROM games WHERE id = 3`.

### PR DESCRIPTION
**Pull request by adngdb**
@cvan r?

Do we want to publicly expose the UUID?

### CODE REVIEW

File: api/models/game.js [hunk 1]
```diff
@@ -84,9 +91,9 @@ Game.objects.all = function (data) {
 Game.objects.create = function (data) {
   return new Promise(function (resolve, reject) {
     db.query(
-      'INSERT INTO games (slug, game_url, name, description, created) ' +
-      'VALUES ($1, $2, $3, $4, NOW()) RETURNING *',
-      [data.slug, data.game_url, data.name, data.description],
+      'INSERT INTO games (uuid, slug, game_url, name, description, created) ' +
+      'VALUES ($1, $2, $3, $4, $5, NOW()) RETURNING *',
+      [UUID.v4(), data.slug, data.game_url, data.name, data.description],
```
Comment (cvan) [inline api/models/game.js]: evidently, [postgres has a UUID type](http://www.postgresql.org/docs/8.3/static/datatype-uuid.html). could we use that?

File: migrations/1420816491290_addUuidToGames.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+exports.up = function(pgm, run) {
+  pgm.addColumns('games', {
+    uuid: {type: 'string', unique: true}
+  })
```
Comment (cvan) [inline migrations/1420816491290_addUuidToGames.js]: missing a semicolon

File: migrations/1420816491290_addUuidToGames.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+exports.up = function(pgm, run) {
```
Comment (cvan) [inline migrations/1420816491290_addUuidToGames.js]: can you change `function(` to be `function (`?

**Comment by cvan** [comment]
> Do we want to publicly expose the UUID?

if we're going to be truly REST-ful, no (since it's not an updatable field). all fields in `POST` and `PUT` (and in `GET` responses) should correspond to columns that are updatable in the DB.

**Comment by cvan** [comment]
looking good. thanks for working on this!

**Comment by cvan** [comment]
do you need to change any existing tests or add new ones?

**Comment by AdrianGaudebert** [comment]
I didn't even think Postgres could have a ``uuid`` type, heh. Thanks for letting me know. 

Regarding unit tests, I haven't added one because the only valid test I can think of would be to pull the data from the database and verify there's something that looks like a uuid in the uuid field. That seems a bit trivial for a unit test, and what's worse: I even failed doing it for reasons I don't understand. If you think that's needed, I'll put more effort into that next time I work on galaxy. :)

**Comment by cvan** [comment]
I think a direct test probably isn't necessary. the migrations must be working because the tests would fail at database creation and schema migrations.

I've got local tests for `GET /games/{idOrSlug}` (#302), so that will kind of handle testing this.

<hr>

one thing I realised: I know a UUID collision is (nearly?) impossible (AFAIK), but could we handle that gracefully in the controller?

File: api/models/game.js [hunk 4]
```diff
@@ -144,7 +151,7 @@ Game.objects.remove = function (data) {
 
     return new Promise(function (resolve, reject) {
       var query = (utils.isStringAnInt(data.idOrSlug) ?
```
Comment (cvan) [inline api/models/game.js]: we may want to rename `idOrSlug` to `uuidOrSlug` everywhere.

**Comment by cvan** [comment]
> > Do we want to publicly expose the UUID?
> 
> if we're going to be truly REST-ful, no (since it's not an updatable field). all fields in `POST` and `PUT` (and in `GET` responses) should correspond to columns that are updatable in the DB.

I just started thinking about this and how would the game developers ever know the UUID?
we'd probably need to change the `Location` header when a game is created:

https://github.com/AdrianGaudebert/galaxy-api/blob/308-uuids/api/models/game.js#L110
https://github.com/AdrianGaudebert/galaxy-api/blob/308-uuids/api/routes.js#L60

do you think we should expose UUID in API responses? what are your thoughts/concerns?

File: api/models/game.js (hunk 1 above)
Comment (cvan) [inline api/models/game.js]: looks like you can do [this](http://stackoverflow.com/a/21327318):

```sql
galaxy-api=# SELECT uuid_in(md5(now()::text)::cstring);
               uuid_in
--------------------------------------
 34bd1352-c185-d7fb-29ea-37c26cc8e8b5
(1 row)
```

**Comment by AdrianGaudebert** [comment]
> one thing I realised: I know a UUID collision is (nearly?) impossible (AFAIK), but could we handle that gracefully in the controller?

I think that's not a good idea. Collisions are indeed theoretically possible, but in reality very unlikely, and the only effect it would have would be that the creation of a game would fail. The upside would be security for a nearly impossible case, the downsides would be that we add almost useless code (and thus complexity), and I don't see how to test that without horribly hacking the code (and I think not having a test for such a rare case would the same as not supporting that case). 

tl;dr: let's not do that. :)

**Comment by cvan** [comment]
okay, as long as the response is a generic 500 and no error leaks. sounds good to me.

**Comment by AdrianGaudebert** [comment]
I like the idea of returning the UUID in the Location header when the resource is created. I suppose it's the responsibility of the client to store that. 

I still don't fully understand the reason why we have both an ID and a slug though, so if you have an explanation for that it would help me get a clearer picture here.

**Comment by cvan** [comment]
> I still don't fully understand the reason why we have both an ID and a slug though, so if you have an explanation for that it would help me get a clearer picture here.

https://galaxy.dev.mozaws.net/games/pokemon/ will be the URL for a game, which will fetch from https://api-galaxy.dev.mozaws.net/games/pokemon/ – I thought it would be nice to be able to look up by slug. thoughts?

**Comment by AdrianGaudebert** [comment]
Oh yeah looking up by slug is a nice feature, I was just wondering what was the point of having both slug and uuid.

File: api/models/game.js (hunk 1 above)
Comment (AdrianGaudebert) [inline api/models/game.js]: This generates a UUID v3 (or v1, I'm not entirely sure, but it's definitely not a v4) and we should be using UUID v4 here. See http://stackoverflow.com/questions/20342058/which-uuid-version-to-use. So we can either use Postgres' ``uuid-ossp`` extension, or keep generating the UUID in the code base. It seems easier to install a node package than a Postgres extension, so I'd be in favor of ``node-uuid``. What do you think?

File: test/api/controllers/game.js [hunk 5]
```diff
@@ -37,15 +37,32 @@ function submitGame(done) {
...
 
       var Game = require('../../../api/models/game');
-      Game.objects.exists({idOrSlug: internals.sampleGameObj.slug})
+      Game.objects.exists({uuidOrSlug: internals.sampleGameObj.slug})
       .then(function (result) {
 
         Code.expect(result).to.equal(true);
 
-        return Game.objects.get({idOrSlug: internals.sampleGameObj.slug});
+        return Game.objects.exists({uuidOrSlug: uuid});
```
Comment (AdrianGaudebert) [inline test/api/controllers/game.js]: Yay! We are testing uuids!

Comment (cvan) [inline test/api/controllers/game.js]: heh, nice, but now we're not testing slugs right? lol

File: api/models/game.js (hunk 1 above)
Comment (cvan) [inline api/models/game.js]: yeah, totally - just stick with with the `node-uuid` package.

File: test/api/controllers/game.js (hunk 5 above)
Comment (AdrianGaudebert) [inline test/api/controllers/game.js]: Oh, yes we are, slugs are being tested at line 54. I simply duplicated the ``exists`` test, and the duplicate uses the UUID instead of the slug.

Comment (cvan) [inline test/api/controllers/game.js]: ah, thanks

**Comment by AdrianGaudebert** [comment]
@cvan Rebased and addressed a few issues caused by recent commits. How do you like it now?

File: api/controllers/game.js [hunk 6]
```diff
@@ -88,7 +88,7 @@ exports.update = {
       var body = Game.getPublicObj(res);
 
       if (res._slugChanged) {
-        reply(body).redirect('/games/' + body.slug);
+        reply(body).redirect('/games/' + res._uuid);
```
Comment (cvan) [inline api/controllers/game.js]: can't you just use `res.uuid`?

File: api/models/game.js [hunk 7]
```diff
@@ -185,6 +193,7 @@ Game.objects.update = function (dataToFetchBy, dataToSave) {
 
         if (game.slug !== res.slug) {
           res._slugChanged = true;
+          res._uuid = game.uuid;
```
Comment (cvan) [inline api/models/game.js]: the UUID isn't changing, so why do you need to set this?

File: test/lib/utils/index.js [hunk 8]
```diff
@@ -49,41 +49,43 @@ lab.experiment('utils.errors', function () {
...
-    done();
-  });
-
-  lab.test('returns true for negative integers', function (done) {
-
-    Code.expect(utils.isStringAnInt('5')).to.equal(true);
+    // UUID v1
+    Code.expect(utils.isStringAUuid('e5d453c0-9ca9-11e4-89d3-123b93f75cba'))
+        .to.equal(false);
+    // UUID v3
```
Comment (cvan) [inline test/lib/utils/index.js]: good tests

File: test/api/controllers/game.js [hunk 9]
```diff
@@ -37,15 +42,27 @@ function submitGame(done) {
       Code.expect(res.result).to.deep.equal(internals.sampleGameObj);
 
       Code.expect(res.statusCode).to.equal(201);
-      Code.expect(res.headers.location).to.equal('/games/no-flex-zone');
+
+      // Make sure the Game has a valid UUID v4.
+      Code.expect(res.headers.location).to.match(gameUrlRegex);
```
Comment (cvan) [inline test/api/controllers/game.js]: I guess this is okay, but feels like a verbose test

File: test/api/controllers/game.js [hunk 10]
```diff
@@ -37,15 +42,27 @@ function submitGame(done) {
...
+
+      var uuid = res.headers.location.match(new RegExp(
+        '[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}'
```
Comment (cvan) [inline test/api/controllers/game.js]: can we just split on the URL? like

```js
res.headers.location.substring(res.headers.location.lastIndexOf('/') + 1)
```

File: test/api/controllers/game.js [hunk 11]
```diff
@@ -285,12 +302,12 @@ lab.experiment('game delete', function () {
...
         var Game = require('../../../api/models/game');
-        Game.objects.exists({idOrSlug: 'no-flex-zone'}).catch(function (err) {
-
-          Code.expect(err).to.contain({name: 'DoesNotExist'});
-
-          done();
-        });
+        Game.objects.exists({uuidOrSlug: 'no-flex-zone'}).catch(
+          function (err) {
+            Code.expect(err).to.contain({name: 'DoesNotExist'});
```
Comment (cvan) [inline test/api/controllers/game.js]: newline?

File: test/api/controllers/game.js [hunk 12]
```diff
@@ -437,7 +454,7 @@ lab.experiment('game update', function () {
         Code.expect(res.result).to.deep.equal(updatedGameObj);
 
         Code.expect(res.statusCode).to.equal(302);
-        Code.expect(res.headers.location).to.equal('/games/throw-some-mo');
+        Code.expect(res.headers.location).to.match(gameUrlRegex);
```
Comment (cvan) [inline test/api/controllers/game.js]: this feels weird, but OK

Comment (cvan) [inline test/api/controllers/game.js]: regex assertions just make me uncomfortable

File: test/api/controllers/game.js (hunk 9 above)
Comment (cvan) [inline test/api/controllers/game.js]: it feels like you should split on the last `/` and just rely on the `utils.isStringAUuid` function (since that's already unit-tested). thoughts?
